src/app/icd/client/CheckInHandler.cpp

```diff
@@ -0,0 +1,191 @@
+    CounterType receivedCheckInCounterOffset = (counter - clientInfo.start_icd_counter) % kCheckInCounterMax;
+
+    // Detect duplicate check-in messages and return CHIP_NO_ERROR on receiving a duplicate message
+    if (receivedCheckInCounterOffset <= clientInfo.offset)
+    {
+        ChipLogError(ICD, "A duplicate check-in message was received and discarded");
+        return CHIP_NO_ERROR;
+    }
+
+    clientInfo.offset = receivedCheckInCounterOffset;
+    bool refreshKey   = (receivedCheckInCounterOffset > kKeyRefreshLimit);
+
+    if (refreshKey)
+    {
+        uint8_t newKeyData[chip::Crypto::kAES_CCM128_Key_Length];
+        mpCheckInDelegate->OnRefreshKeyGenerate(clientInfo, newKeyData, chip::Crypto::kAES_CCM128_Key_Length);
+        // A new session should be established to re-register the client using the new key. The registration will happen in
+        // mOnDeviceConnected callback
+        EstablishSessionToPeer(clientInfo.peer_node);
+    }
+    else
+    {
+        mpCheckInDelegate->OnCheckInComplete(clientInfo);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CheckInHandler::RegisterClientWithNewKey(ICDClientInfo & clientInfo, ByteSpan newKey,
+                                                    Messaging::ExchangeManager & exchangeMgr, const SessionHandle & sessionHandle)
+{
+    // using namespace Clusters::IcdManagement;
+    // TODO : Determine if using an Objectpool of commandSenders is the best approach here
+    //  app::CommandSender registerCommandSender(&registerCommandSenderDelegate, &exchangeMgr);
+
+    // auto commandPathParams = CommandPathParams(0, 0, Id, Commands::RegisterClient::Id, (CommandPathFlags::kEndpointIdValid));
+    // ReturnErrorOnFailure(registerCommandSender.PrepareCommand(commandPathParams));
+
+    // chip::TLV::TLVWriter * writer = registerCommandSender.GetCommandDataIBTLVWriter();
+
+    // ReturnErrorOnFailure(
+    //     writer->Put(chip::TLV::ContextTag(Commands::RegisterClient::Fields::kCheckInNodeID), clientInfo.peer_node.GetNodeId()));
+    // ReturnErrorOnFailure(
+    //     writer->Put(chip::TLV::ContextTag(Commands::RegisterClient::Fields::kMonitoredSubject), clientInfo.monitored_subject));
+    // ReturnErrorOnFailure(writer->Put(chip::TLV::ContextTag(Commands::RegisterClient::Fields::kKey), newKey));
+
+    // // TODO : We don't have plain data for the old key
+
+    // ReturnErrorOnFailure(registerCommandSender.FinishCommand());
+    // ReturnErrorOnFailure(registerCommandSender.SendCommandRequest(sessionHandle));
+
+    return CHIP_NO_ERROR;
+}
+
+void CheckInHandler::EstablishSessionToPeer(ScopedNodeId peerId)
+{
+    ChipLogProgress(ICD, "Trying to establish a CASE session for re-registering an ICD client");
+    auto * caseSessionManager = InteractionModelEngine::GetInstance()->GetCASESessionManager();
+    VerifyOrReturn(caseSessionManager != nullptr);
+    caseSessionManager->FindOrEstablishSession(peerId, &mOnConnectedCallback, &mOnConnectionFailureCallback);
```

Comment:
This is broken.  As soon as we get two separate check-in messages that both need key rollover at the same time, they will stomp on each other right here, no?

Reply:
Discussed offline. Will implement the new design in a folow-up PR.